### PR TITLE
Db migration guide

### PR DESCRIPTION
**Pull request by juandiegopalomino**

### CODE REVIEW

File: content/en/Features/db_migration.md [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+source $(pipenv --venv)/bin/activate
+
+cd srv
+flask db upgrade # Does the migrations
+gunicorn -c gunicorn.py wsgi # Starts the server
+```
+
+As long as the migrations are short, this option should work fine, keep your database and server changes in lock step 
+and keep your ci/cd unchanged.
```
Comment (RemyDeWolf) [inline content/en/Features/db_migration.md:104]: the potential issue with this solution is that the probe check might fail and the pod could be killed during migration. 
We can document this risk a bit more and advise the user to set the `readiness_probe_path` to allow enough time for the migration to complete.

File: content/en/Features/db_migration.md [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    path: "./staging.yaml"
+name: my-service-migration
+modules:
+  - type: external-state
+    name: external
+    backend_type: s3
+    config:
+      bucket: opta-tf-state-my-org-staging
+      key: my-service
+      region: us-east-1
```
Comment (RemyDeWolf) [inline content/en/Features/db_migration.md:135]: why are we using `external-state` here? it's not related to this topic.
Specially since it's a new module, some users might not understand that this is not needed.

File: content/en/Features/db_migration.md [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+                valueFrom:
+                  secretKeyRef:
+                    name: secret
+                    key: DBUSER
+                    optional: true
+            command:
+              - ./home/app/do_migration.sh
+```
+{{< /tab >}}
+{{< tab tabName="Same Manifest" >}}
```
Comment (RemyDeWolf) [inline content/en/Features/db_migration.md:179]: this solution seems simpler, having everything define in the same place makes it easy to understand.
why not only keep this one? or at least place it first.
the other looks more advanced

File: content/en/Features/db_migration.md [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+
+Even in small datasets, database migrations will always face the danger of improper data transformations which cause errors.
+One of the best countermeasures for such errors is testing the migrations in non-production databases which closely
+resemble production-- the more similar the better as it will test more potential edge cases. Typically, this is done
+as part of the deployment to a "staging/qa" environment.
+
+### Have Folks on Standby
+
+Always prepare under the assumption that if something can go wrong, it will go wrong. Hence, always have someone on 
+standby monitoring the migration.
```
Comment (RemyDeWolf) [inline content/en/Features/db_migration.md:78]: this is a good point, but this is also very common to run DB migrations when there is low traffic - so typically at night. 
So the end goal is to have a rollout mechanism and more testing vs having people in standby for each migration. May be for unusual large migrations, have folks in standby makes more sense.

**Review by RemyDeWolf — COMMENTED** [review]
this is really great content!

File: content/en/Features/db_migration.md [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+                    optional: true
+            command:
+              - ./home/app/do_migration.sh
+```
+
+{{< /tab >}}
+{{< /tabs >}}
+
+This approach, while more explicit, does require the user to have knowledge of Kubernetes jobs, and to manually
+connect the required database secrets from the secret set up for the k8s service link (**DO NOT** reference the db creds
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md:247]: they're mounted as secrets, but I'll clarify

File: content/en/Features/db_migration.md [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+
+As long as the migrations are short, this option should work fine, keep your database and server changes in lock step 
+and keep your ci/cd unchanged.
+
+### In a Running Server's Container
+An alternative route is having a user enter one of the running containers via the `opta shell` command and execute the 
+migration manually. This option offers more developer control, but now requires a human in the loop, does not enforce
+synchronicity between the database and server versions, and users will need to be careful not to trigger a new deployment
+while the operation is taking place.
+
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md:111]: `The application might become unstable if the DB schema is changed on the fly.` that will always happen because one way or another servers are actively running when a migration is taking place (unless you wish to bring the whole webapp down while this is happening which no one does).

File: content/en/Features/db_migration.md (hunk 2 above)
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md:135]: `This can be done in either the existing service's manifest, or a new manifest which refers to the current manifest for data such as the current container image deploy`

File: content/en/Features/db_migration.md [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+                    key: DBPASS
+                    optional: true
+              - name: DBUSER
+                valueFrom:
+                  secretKeyRef:
+                    name: secret
+                    key: DBUSER
+                    optional: true
+            command:
+              - ./home/app/do_migration.sh
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: done

File: content/en/Features/db_migration.md [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+  - type: external-state
+    name: external
+    backend_type: s3
+    config:
+      bucket: opta-tf-state-my-org-staging
+      key: my-service
+      region: us-east-1
+  - type: helm-chart
+    name: dbmigration1
+    namespace: my-service
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: done

File: content/en/Features/db_migration.md (hunk 1 above)
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: done

File: content/en/Features/db_migration.md [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+### Understand when Database Migrations are Needed
+Simply put, not all production-grade databases need database migrations. The concept of database migrations have 
+historically been tied to relational databases, and other database models (e.g. key-value, document-store, graphs)
+may offer easier alternatives. For instance, Kubernetes (which uses the etcd key-value database), has mostly bypassed
+database migration issues with its concept of api versioning, where multiple "schema versions" can be supported 
+simultaneously. Protocol buffers offer a similar alternative with its versioning system, allowing 
+[backwards and forward compatibility](https://stackoverflow.com/questions/8519381/how-does-protocol-buffer-handle-versioning).
+
+If you are using one of these non-relational databases, then consider what alternatives you have at your disposal, before
+fully committing. Even if you are using relational databases, many like postgres and mysql have been capable of storing
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: done

File: content/en/Features/db_migration.md [hunk 10]
```diff
@@ -0,0 +1,257 @@
+---
+title: "DB Migrations in Opta"
+linkTitle: "Opta DB Migrations"
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: done

File: content/en/Features/db_migration.md (hunk 3 above)
Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: The job needs to be renamed on every run, which can cause unpleasant complications/errors on regular deploys, which is why I think it ought to be a separate service

Comment (juandiegopalomino) [reply content/en/Features/db_migration.md]: I placed it first now

**Review by RemyDeWolf — APPROVED** [review]
minor comments below
